src/java/org/apache/cassandra/security/AbstractCryptoProvider.java

```diff
@@ -0,0 +1,144 @@
+import org.slf4j.LoggerFactory;
+
+import org.apache.cassandra.exceptions.ConfigurationException;
+
+import static java.lang.String.format;
+
+public abstract class AbstractCryptoProvider
+{
+    protected static final Logger logger = LoggerFactory.getLogger(AbstractCryptoProvider.class);
+    public static final String FAIL_ON_MISSING_PROVIDER_KEY = "fail_on_missing_provider";
+
+    protected final boolean failOnMissingProvider;
+
+    public AbstractCryptoProvider(Map<String, String> properties)
+    {
+        failOnMissingProvider = properties != null && Boolean.parseBoolean(properties.getOrDefault(FAIL_ON_MISSING_PROVIDER_KEY, "false"));
+    }
+
+    /**
+     * Returns name of the provider, as returned from {@link Provider#getName()}
+     *
+     * @return name of the provider
+     */
+    public abstract String getProviderName();
+
+    /**
+     * Returns the name of the class which installs specific provider of name {@link #getProviderName()}.
+     *
+     * @return name of class of provider
+     */
+    public abstract String getProviderClassAsString();
+
+    /**
+     * Returns a runnable which installs this crypto provider.
+     *
+     * @return runnable which installs this provider
+     */
+    protected abstract Runnable installator() throws Exception;
+
+    /**
+     * Returns boolean telling if this provider was installed properly.
+     *
+     * @return {@code true} if provider was installed properly, {@code false} otherwise.
+     */
+    protected abstract boolean isHealthyInstallation() throws Exception;
+
+    /**
+     * The default installation runs {@link AbstractCryptoProvider#installator()} and after that
+     * {@link AbstractCryptoProvider#isHealthyInstallation()}.
+     * <p>
+     * If any step fails, it will not throw an exception unless the parameter
+     * {@link AbstractCryptoProvider#FAIL_ON_MISSING_PROVIDER_KEY} is {@code true}.
+     */
+    public void install() throws Exception
+    {
+        if (NoOpCryptoProvider.class.getName().equals(getProviderClassAsString()))
+        {
+            logger.debug("Installation of a crypto provider was skipped.");
+            return;
+        }
+
+        String failureMessage = null;
+        try
+        {
+            Class.forName(getProviderClassAsString());
+
+            Provider[] providers = Security.getProviders();
+
+            if (providers.length > 0 && providers[0] != null && providers[0].getName().equals(getProviderName()))
+            {
+                logger.debug("{} was already installed", getProviderName());
+                return;
+            }
+            else
+            {
+                installator().run();
+            }
+
+            if (isHealthyInstallation())
+            {
+                logger.info("{} successfully passed the healthiness check", getProviderName());
+            }
+            else
+            {
+                failureMessage = format("%s is not the highest priority provider. " +
```

Comment:
is this the situation where the fallback to the JRE crypto provider happens?

Reply:
Yes. And that happens if `fail_on_missing_provider` is `false`, if `true`, we fail the startup.

Reply:
The assumption is that if it does not pass the healthcheck, it is not installed. So the provider which was there (the default one) has not changed, so the "fallback" is kind of automatic. We are not returning to anything per se.

Reply:
https://github.com/corretto/amazon-corretto-crypto-provider#verification-optional

Reply:
I hardened this logic a bit. We may uninstall the provider in case it was installed but healthcheck has failed. There might be other crypto providers which do it differently than Corretto.